### ISSUE TITLE
Flag records that become invalid with respect to their current workflow state

### ISSUE THREAD

**Issue by mark-saeon** [SAEONData/ckanext-metadata#38]
In a similar way to how we automatically invalidate metadata records following specific change scenarios, we must deal with cases in which a metadata record in some workflow state (and which has therefore undergone workflow rules validation to enter that state) would now no longer be valid for that state if workflow validation were run again. This might occur if, for example, the JSON schema defining the rules for that state is changed.

The problem is that unlike with metadata invalidation - which is easy to automate since there are a specific set of internal triggers and it won't lead to any inconsistencies in the database - workflow validation might have external dependencies (e.g. vocabulary service lookups), so we cannot necessarily trigger automatic flagging of records in all scenarios in which workflow rules might suddenly be violated. The best approach would be to run a background process which periodically validates a record against the rules for the workflow state that it is currently in, and flag the record if that validation fails.

**Comment by mark-saeon** [comment]
Hmm. Metadata validation might also have external dependencies (vocab lookups), so we need to also consider a background process for periodically re-validating metadata - in addition to the automatic invalidation already in place. Consider, however, that vocab dependencies might be linked explicitly with metadata records (see #26), and if internal vocabularies are automatically synced with their external dependencies (see #29) then we have the automatic trigger we need.

The bottom line is that we must explicitly declare all scenarios in which either metadata invalidation or workflow flagging are required, and determine whether - in each case - automatic triggering is possible. If there are any cases in which we cannot devise a mechanism to respond automatically to some external change that requires re-validation of either metadata or workflow rules, then a background process is probably necessary.
